File: java/gpt/p143_nre.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import javax.sql.DataSource;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class p143_nre {

  private static final String GENERIC_MSG = "If this email was subscribed, it has been unsubscribed.";
  private static final Pattern EMAIL_RE = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");
  private final JdbcTemplate jdbc;
  private final byte[] secret;

  public p143_nre(DataSource ds) {
    this.jdbc = new JdbcTemplate(ds);
    this.secret = System.getenv().getOrDefault("UNSUB_SECRET", "replace-with-strong-secret")
        .getBytes(StandardCharsets.UTF_8);
  }

  private static String normalizeEmail(String raw) {
    if (raw == null) return null;
    String s = raw.trim();
    if (s.length() < 1 || s.length() > 254) return null;
    if (!EMAIL_RE.matcher(s).matches()) return null;
    int at = s.lastIndexOf('@');
    if (at <= 0 || at >= s.length() - 3) return null;
    String local = s.substring(0, at);
    String domain = s.substring(at + 1).toLowerCase();
    return local + "@" + domain;
  }

  private static byte[] b64urlDecode(String s) {
    String str = s.replace('-', '+').replace('_', '/');
    int pad = (4 - (str.length() % 4)) % 4;
    str = str + "=".repeat(pad);
    return java.util.Base64.getDecoder().decode(str);
  }

  private static boolean constantTimeEquals(byte[] a, byte[] b) {
    return MessageDigest.isEqual(a, b);
  }

  private String verifyToken(String token) {
    try {
      String[] parts = token.split("\\.", 2);
      if (parts.length != 2) return null;

      byte[] payload = b64urlDecode(parts[0]);
      byte[] sig = b64urlDecode(parts[1]);

      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(new SecretKeySpec(secret, "HmacSHA256"));
      byte[] expected = mac.doFinal(payload);

      if (sig.length != expected.length || !constantTimeEquals(sig, expected)) return null;

      Map<String, String> kv = parseQuery(new String(payload, StandardCharsets.UTF_8));
      long exp = Long.parseLong(kv.getOrDefault("exp", "0"));
      if (exp <= Instant.now().getEpochSecond()) return null;

      String email = normalizeEmail(kv.get("email"));
      if (email == null) return null;

      // Optional: enforce nonce single-use: kv.get("nonce")
      return email;
    } catch (Exception e) {
      return null;
    }
  }

  private static Map<String, String> parseQuery(String qs) {
    Map<String, String> m = new HashMap<>();
    String[] parts = qs.split("&");
    for (String p : parts) {
      int i = p.indexOf('=');
      if (i > 0 && i < p.length() - 1) {
        m.put(p.substring(0, i), p.substring(i + 1));
      }
    }
    return m;
  }

  private void mailinglistUnsubscribe(String email) {
    // Idempotent external call placeholder
  }

  @GetMapping(value = "/unsubscribe", produces = MediaType.TEXT_PLAIN_VALUE)
  public ResponseEntity<String> unsubscribe(@RequestParam(name = "token", required = false) String token,
                                            HttpServletRequest req) {
    String email = verifyToken(token == null ? "" : token);

    if (email != null) {
      try {
        jdbc.update("DELETE FROM users WHERE email = ?", email); // parameterized
        mailinglistUnsubscribe(email);
      } catch (Exception ignored) {
      }
    }

    return ResponseEntity.ok(GENERIC_MSG);
  }
}

File: javascript/gpt/p143_nre.js
import express from "express";
import crypto from "crypto";
import { Pool } from "pg";

const app = express();

const GENERIC_MSG = "If this email was subscribed, it has been unsubscribed.";
const SECRET = Buffer.from(process.env.UNSUB_SECRET || "replace-with-strong-secret", "utf8");

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
});

function normalizeEmail(raw) {
  if (typeof raw !== "string") return null;
  const s = raw.trim();
  if (s.length < 1 || s.length > 254) return null;
  const ok = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(s);
  if (!ok) return null;
  const at = s.lastIndexOf("@");
  const local = s.slice(0, at);
  const domain = s.slice(at + 1).toLowerCase();
  return `${local}@${domain}`;
}

function b64urlDecode(str) {
  const pad = "=".repeat((4 - (str.length % 4)) % 4);
  return Buffer.from(str.replace(/-/g, "+").replace(/_/g, "/") + pad, "base64");
}

function timingSafeEqual(a, b) {
  if (a.length !== b.length) return false;
  return crypto.timingSafeEqual(a, b);
}

function verifyToken(token) {
  try {
    const [p1, p2] = token.split(".", 2);
    if (!p1 || !p2) return null;

    const payload = b64urlDecode(p1);
    const sig = b64urlDecode(p2);

    const expected = crypto.createHmac("sha256", SECRET).update(payload).digest();
    if (!timingSafeEqual(sig, expected)) return null;

    const qs = new URLSearchParams(payload.toString("utf8"));
    const exp = Number(qs.get("exp") || "0");
    if (!Number.isFinite(exp) || exp <= Math.floor(Date.now() / 1000)) return null;

    const email = normalizeEmail(qs.get("email") || "");
    if (!email) return null;

    // Optional: enforce nonce single-use in DB/cache: qs.get("nonce")
    return email;
  } catch {
    return null;
  }
}

async function mailinglistUnsubscribe(email) {
  // Idempotent external call placeholder
  void email;
}

app.get("/unsubscribe", async (req, res) => {
  const token = String(req.query.token || "");
  const email = verifyToken(token);

  if (!email) {
    res.status(200).type("text/plain").send(GENERIC_MSG);
    return;
  }

  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    await client.query("DELETE FROM users WHERE email = $1", [email]); // parameterized
    await client.query("COMMIT");
  } catch {
    try { await client.query("ROLLBACK"); } catch {}
  } finally {
    client.release();
  }

  try { await mailinglistUnsubscribe(email); } catch {}

  res.status(200).type("text/plain").send(GENERIC_MSG);
});

app.listen(process.env.PORT || 3000);